xdsl/backend/csl/lowering/convert_stencil_to_csl.py

```diff
@@ -0,0 +1,399 @@
+    layout.add_op(one_u := SignednessCastOp.get_u(one))
+    layout.add_op(width_sl := SignednessCastOp.get(width))
+    layout.add_op(height_sl := SignednessCastOp.get(height))
+
+    # setting up two-dimensional loop nest over physical x-y PEs
+    outer_loop_body = Block(arg_types=[IntegerType(16)])
+    inner_loop_body = Block(arg_types=[IntegerType(16)])
+
+    # adding outer loop to layout block
+    layout.add_op(
+        For(lb=zero, ub=width_sl, step=one, iter_args=[], body=outer_loop_body)
+    )
+
+    # signedness cast for outer loop variable
+    # outer loop is x_id
+    outer_loop_body.add_op(x_id := SignednessCastOp.get_u(outer_loop_body.args[0]))
+
+    # adding inner loop inside outer loop
+    outer_loop_body.add_op(
+        For(lb=zero, ub=height_sl, step=one, iter_args=[], body=inner_loop_body)
+    )
+
+    # preparing calls to route module, memcpy module, and `@set_tile_code`
+    inner_loop_body.add_ops(
+        [
+            # inner loop is y_id
+            y_id := SignednessCastOp.get_u(inner_loop_body.args[0]),
+            # compute boolean expression is_border_region_pe
+            pattern_minus_one := MinUI(pattern_op, one_u),
+            width_minus_xid := MinUI(width, x_id),
+            height_minus_yid := MinUI(height, y_id),
+            first := Cmpi(x_id, pattern_minus_one, "ult"),
+            second := Cmpi(y_id, pattern_minus_one, "ult"),
+            third := Cmpi(width_minus_xid, pattern_op, "ult"),
+            fourth := Cmpi(height_minus_yid, pattern_op, "ult"),
+            or_one := OrI(first, second),
+            or_two := OrI(or_one, third),
+            is_border_pe := OrI(or_two, fourth),
+            # generates: memcpy.get_params(xId)
+            memcpy_params := MemberCallOp.get(
+                "get_params", ComptimeStructType(), memcpy, x_id
+            ),
+            # generates: routes.computeAllRoutes(x_id, y_id, width, height, pattern)
+            route_params := MemberCallOp.get(
+                "computeAllRoutes",
+                ComptimeStructType(),
+                routes,
+                x_id,
+                y_id,
+                width,
+                height,
+                pattern_op,
+            ),
+            # setting up param structs for `@set_tile_code`
+            set_tile_params := ConstStructOp.get(
+                (Named.memcpy_params, memcpy_params),
+                (Named.stencil_comms_params, route_params),
+            ),
+            params_task := ConstStructOp.get((Named.is_border_region_pe, is_border_pe)),
+            set_tile_params_ext := ConcatStructOp.get(params_task, set_tile_params),
+            # generates: @set_tile_code(xId, yId, "pe.csl", .. param structs ..)
+            SetTileCodeOp.get(
+                ctx.program_module.sym_name,
+                x_id,
+                y_id,
+                set_tile_params_ext,
+            ),
+            Yield(),
+        ]
+    )
+    outer_loop_body.add_op(Yield())
```

Comment:
I would highly suggest you look into `ImplicitBuilder`, which is our attempt at making IR construction readable :)

Reply:
thanks, will take a look. It seems mostly used for testing though?

Reply:
This is actually quite neat, using the ImplicitBuilder now